fastiron/src/constants.rs

```diff
@@ -45,105 +45,116 @@ pub type Tuple4 = (usize, usize, usize, usize);
 
 // generic float type
 
+/// Associated reference value used for compute approximation.
+///
+/// This is the only trait that should be manually implemented for custom types. The `CustomFloat` trait is then
+/// automatically implemented via a blanket implementation.
+pub trait CustomReferenceFloat: Float + FromPrimitive {
+    /// Threshold upper-value for decimal number.
+    const HUGE_FLOAT: Self;
+    /// Threshold low-ish-value for decimal number.
+    const SMALL_FLOAT: Self;
+    /// Threshold lower-value for decimal number.
+    const TINY_FLOAT: Self;
+    /// Pi value.
+    const PI: Self;
+}
+
 /// Custom trait for floating point number
-pub trait OpsFloat: AddAssign + SubAssign + MulAssign + DivAssign + Sized {}
-/// Custom trait for floating point number
-pub trait UtilsFloat:
-    Default
+pub trait CustomFloat:
+    Float
+    + CustomReferenceFloat
+    + Default
+    // conversions
+    + FromPrimitive
+    + FromStr
+    // operations
+    + AddAssign
+    + SubAssign
+    + MulAssign
+    + DivAssign
+    + Sum
+    // display
     + Debug
     + Display
     + LowerExp
-    + FromStr
-    + From<f32>
-    + Sum
-    + for<'a> Sum<&'a Self>
+    // parallel-safe
     + Send
     + Sync
 {
-}
-/// Custom super-trait for floating point number
-pub trait CustomFloat: Float + FromPrimitive + OpsFloat + UtilsFloat {
-    // floating ref
-
     /// Threshold upper-value for decimal number.
-    fn huge_float<T: CustomFloat>() -> T;
-    /// Threshold low-ish-value for decimal number.
-    fn small_float<T: CustomFloat>() -> T;
-    /// Threshold lower-value for decimal number.
-    fn tiny_float<T: CustomFloat>() -> T;
-
-    // physical constants
-
-    /// Neutron mass energy equivalent (MeV)
-    fn neutron_mass_energy<T: CustomFloat>() -> T;
-    /// [Pick your definition][3]
-    ///
-    /// [3]: https://en.wikipedia.org/wiki/Pi
-    fn pi<T: CustomFloat>() -> T;
-    /// Speed of light (cm/s)
-    fn light_speed<T: CustomFloat>() -> T;
-}
-
-impl OpsFloat for f32 {}
-impl UtilsFloat for f32 {}
-impl CustomFloat for f32 {
-    /// Threshold value for decimal number when using [f32]. May need adjustment.
     fn huge_float<T: CustomFloat>() -> T {
-        FromPrimitive::from_f32(10e35_f32).unwrap()
+        T::from(T::HUGE_FLOAT).unwrap()
```

Comment:
Why `from` and `unwrap`. I thought the `T::HUGE_FLOAT` was already a `T`.

Reply:
You're right, `from` & `unwrap` aren't necessary here.